WcaOnRails/app/controllers/application_controller.rb

```diff
@@ -14,8 +14,26 @@ class ApplicationController < ActionController::Base
   end
 
   def set_locale
-    # We set this to have access to the locale in controller
-    I18n.locale = params[:locale] || I18n.default_locale
+    # If the locale for the session is not set, we want to infer it from the following sources:
+    #  - the current user preferred locale
+    #  - the Accept-Language http header
+    session[:locale] ||= current_user&.preferred_locale || http_accept_language.preferred_language_from(I18n.available_locales)
+    I18n.locale = session[:locale] || I18n.default_locale
+  end
+
+  def update_locale
+    # Validate the requested locale by looking at those available
+    if params[:locale] && I18n.available_locales.include?(params[:locale].to_sym)
+      session[:locale] = params[:locale]
+
+      current_user&.update preferred_locale: session[:locale]
```

Comment:
If my understanding of the doc is correct, the main difference between both methods is that the latter raises an exception if the object is invalid. If I'm not missing anything then we can just use the return value of `update` to change the flash message and not bother with an exception using `update!`.
(I've just did so in my last commit)